File: September/2026-09-04/457/457-circular-arrray-loops.py
#  457-circular-arrray-loops.cpp;array-2p;help;leetcode

class Solution:
    def nextIndex(self, nums: List[int], index: int) -> int:
        n = len(nums)

        return ((index + nums[index]) % n + n) % n

    def circularArrayLoop(self, nums: List[int]) -> bool:
        n = len(nums)

        for i in range(n):
            forward = nums[i] > 0

            slow = i
            fast = i

            while True:
                nextSlow = self.nextIndex(nums, slow)

                if (nums[nextSlow] > 0) != forward:
                    break

                if nextSlow == slow:
                    break

                slow = nextSlow

                nextFast = self.nextIndex(nums, fast)

                if (nums[nextFast] > 0) != forward:
                    break

                if nextFast == fast:
                    break

                nextFast = self.nextIndex(nums, nextFast)

                if (nums[nextFast] > 0) != forward:
                    break

                if nextFast == self.nextIndex(nums, fast):
                    break

                fast = nextFast

                if slow == fast:
                    return True

        return False
        

File: September 2026/2026-09-04/457/457-circular-arrray-loops.cpp
// 457-circular-arrray-loops.cpp;array-2p;help;leetcode

class Solution {
public:
    int nextIndex(vector<int>& nums, int index) {
    int n = nums.size();

    return ((index + nums[index]) % n + n) % n;
}

bool circularArrayLoop(vector<int>& nums) {
    int n = nums.size();

    for (int i = 0; i < n; i++) {

        bool forward = nums[i] > 0;

        int slow = i;
        int fast = i;

        while (true) {
            int nextSlow = nextIndex(nums, slow);

            if ((nums[nextSlow] > 0) != forward)
                break;

            if (nextSlow == slow)
                break;

            slow = nextSlow;

            int nextFast = nextIndex(nums, fast);

            if ((nums[nextFast] > 0) != forward)
                break;

            if (nextFast == fast)
                break;

            nextFast = nextIndex(nums, nextFast);

            if ((nums[nextFast] > 0) != forward)
                break;

            if (nextFast == nextIndex(nums, fast))
                break;

            fast = nextFast;

            if (slow == fast)
                return true;
        }
    }

    return false;
}
};